### PR TITLE
[WIP] FSMT bart-like refactor

### PR DESCRIPTION
**Pull request by patil-suraj**
# What does this PR do?

This PR refactors `FSMT` to align it with other (bart-like) seq-2-seq models in the lib.

This PR refactors `FSMT` similar to `Bart` in that it moves the time dimension to be always at the 2nd place and the batch dimensions always in the first place. Also, the cache is refactored to consists of `tuples` instead of a `dict`. 

This refactor is very similar to  #10501.

I have verified that all slow-tets are passing and that all metrics (BLEU score) can be reproduced.  I ran the evaluation of the following four models and the results are similar to those reported in the [model cards](https://huggingface.co/facebook/wmt19-en-ru).
- en-ru: 33.42
- ru-en: 39.20
- en-de: 42.83
- de-en: 41.39

### Benchmarking

This PR however introduces some speed and memory regression, which I'm currently investigating.

On this PR:
```
====================       INFERENCE - SPEED - RESULT       ====================
--------------------------------------------------------------------------------
          Model Name             Batch Size     Seq Length     Time in s   
--------------------------------------------------------------------------------
     facebook/wmt19-en-ru            4               8             0.009     
     facebook/wmt19-en-ru            4               32             0.01     
     facebook/wmt19-en-ru            4              128            0.026     
     facebook/wmt19-en-ru            4              512            0.109     
--------------------------------------------------------------------------------

====================      INFERENCE - MEMORY - RESULT       ====================
--------------------------------------------------------------------------------
          Model Name             Batch Size     Seq Length    Memory in MB 
--------------------------------------------------------------------------------
     facebook/wmt19-en-ru            4               8              2172     
     facebook/wmt19-en-ru            4               32             2200     
     facebook/wmt19-en-ru            4              128             2306     
     facebook/wmt19-en-ru            4              512             2792     
--------------------------------------------------------------------------------
```

On master:

```
====================       INFERENCE - SPEED - RESULT       ====================
--------------------------------------------------------------------------------
          Model Name             Batch Size     Seq Length     Time in s   
--------------------------------------------------------------------------------
     facebook/wmt19-en-ru            4               8             0.007     
     facebook/wmt19-en-ru            4               32            0.007     
     facebook/wmt19-en-ru            4              128            0.013     
     facebook/wmt19-en-ru            4              512            0.046     
--------------------------------------------------------------------------------

====================      INFERENCE - MEMORY - RESULT       ====================
--------------------------------------------------------------------------------
          Model Name             Batch Size     Seq Length    Memory in MB 
--------------------------------------------------------------------------------
     facebook/wmt19-en-ru            4               8              2170     
     facebook/wmt19-en-ru            4               32             2176     
     facebook/wmt19-en-ru            4              128             2204     
     facebook/wmt19-en-ru            4              512             2356     
--------------------------------------------------------------------------------


```

### CODE REVIEW

File: src/transformers/models/fsmt/modeling_fsmt.py
Comment (patil-suraj) [inline src/transformers/models/fsmt/modeling_fsmt.py:310]: IMO `decoder_input_ids` should not be created using `input_ids` for this model, because it uses two different vocabs for the encoder and decode and this is really only applicable for BART because it uses the `input_ids` for task like classification and QA which FSMT does not support.

File: utils/check_repo.py
Comment (sgugger) [inline utils/check_repo.py:94]: This should not be needed anymore.

**Review by sgugger — APPROVED** [review]
Thanks a lot for working on this!

**Review by stas00 — APPROVED** [review]
Thank you for doing this refactoring, @patil-suraj!

It's a bit hard to review since all the code is moved around, so no easy diff to follow - so while I skimmed through it - I trust your expertise and the tests on the correctness.

With regards to memory/performance regression - (thank you for running this important check!) could it be that it was introduced in the initial Bart refactor? i.e. perhaps running the same check on Bart pre and post PR that did the main refactoring (when all the Barts were split up)? And if so then the issue is bigger and needs to be looked in that PR that introduced it.

**Comment by patil-suraj** [comment]
Thanks Stas! I'm not sure what exactly introduced this memory/speed regression, so I'm going to investigate it and won't merge this PR before that.

**Comment by patrickvonplaten** [comment]
> Thank you for doing this refactoring, @patil-suraj!
> 
> It's a bit hard to review since all the code is moved around, so no easy diff to follow - so while I skimmed through it - I trust your expertise and the tests on the correctness.
> 
> With regards to memory/performance regression - (thank you for running this important check!) could it be that it was introduced in the initial Bart refactor? i.e. perhaps running the same check on Bart pre and post PR that did the main refactoring (when all the Barts were split up)? And if so then the issue is bigger and needs to be looked in that PR that introduced it.

I remember that I checked that the [Bart refactor](https://github.com/huggingface/transformers/pull/8900) didn't show any regression both on the forward pass and `generate()`. I might however have overlooked something. Would definitely be a good idea to verify this first with the exact same testing params (batch_size=4, ...)!

File: src/transformers/models/fsmt/modeling_fsmt.py
Comment (patrickvonplaten) [inline src/transformers/models/fsmt/modeling_fsmt.py:280]: I'm actually not sure whether using MBart's way of creating the tokens here is correct. Note that MBart uses a very special language-dependent pattern when fine-tuning so that the `decoder_start_token_id` depends on the language, see Fig. 1 of [paper](https://arxiv.org/pdf/2001.08210.pdf). 

IMO, this is not the case for FSMT. FSMT always uses the EOS/SEP token as the `decoder_start_token_id = sep_token_id = "<\s>"` if I understand correctly => Think it's better therefore to copy the `shift_tokens_right` function from Bart and set `decoder_start_token_id = 2` as a default in the config. This would make this function much easier. 

Would be great if you could confirm @stas00

File: src/transformers/models/fsmt/modeling_fsmt.py
Comment (patrickvonplaten) [reply src/transformers/models/fsmt/modeling_fsmt.py:310]: Yes!

File: src/transformers/models/fsmt/modeling_fsmt.py
Comment (patil-suraj) [reply src/transformers/models/fsmt/modeling_fsmt.py:280]: Yes, you are right. This is also equivalent in that, inf FSMT sequence ends with `EOS` and then this method moves it to the beginning. But yeah, I will copy  BARTs  function, which is more explicit here.

File: src/transformers/models/fsmt/modeling_fsmt.py
Comment (patrickvonplaten) [inline src/transformers/models/fsmt/modeling_fsmt.py:757]: maybe rename this to `FSMTPreTrainedModel` and deprecate this naming as it's done for Bart: https://github.com/huggingface/transformers/blob/5a34d8d982515fa8723a5e95a1301c3e48e6806a/src/transformers/models/bart/modeling_bart.py#L504 ?

**Review by patrickvonplaten — COMMENTED** [review]
Thanks a lot for the PR! Great to see much more `Copied from` statements! I'm also very curios where the memory/speed regression comes from. Think it's a good idea to make sure that the original Bart refactor didn't introduce this regression - maybe you can check the exact same testing params with Bart before the [refactor](https://github.com/huggingface/transformers/pull/8900) and after & then see from there.

Another small nit: after this PR you should be able to delete this line in the common tests:
https://github.com/huggingface/transformers/blob/5a34d8d982515fa8723a5e95a1301c3e48e6806a/tests/test_modeling_common.py#L404

File: src/transformers/models/fsmt/modeling_fsmt.py
Comment (stas00) [reply src/transformers/models/fsmt/modeling_fsmt.py:280]: Indeed! Thank you for catching that, @patrickvonplaten!

**Comment by stas00** [comment]
@patil-suraj, what needs to be done to complete this? 

Last we talked there was a performance regression and the suggestion was to test Bart's performance pre and post its original refactoring.

**Comment by stas00** [comment]
@patil-suraj, FYI, recently I made a few fixes to the model to make it work with Deepspeed:
https://github.com/huggingface/transformers/pull/12477/files#diff-564f6d9b78eec17b410c924f868840770a9ad9649032bcf3754827317b9eaba3

Are we still planning to merge this PR? As we said earlier if there is a regression it'll be on the whole Bart family, so perhaps it might be easier to just merge this? Otherwise a lot of time gets waste getting back to it again and again and not getting anywhere.

Thank you.

**Comment by ArthurZucker** [comment]
Closing as this PR is super old

**Pull request closed without merging**
